authorizer-app-backend/src/main/java/org/radarbase/authorizer/service/OAuth1RestSourceAuthorizationService.kt

```diff
@@ -0,0 +1,183 @@
+        val req = createRequest("POST", tokenEndpoint.orEmpty(), tokens, sourceType)
+
+        return httpClient.newCall(req).execute().use { response ->
+            when (response.code) {
+                200 -> response.body?.string()
+                        ?.let { tokenReader.readValue<RestOauth1AccessToken>(parseParams(it)) }
+                        ?: throw HttpBadGatewayException("Service did not provide a result")
+                400, 401, 403 ->{
+                    logger.info(response.body?.string())
+                    null }
+                else -> throw HttpBadGatewayException("Cannot connect to ${tokenEndpoint}: HTTP status ${response.code}")
+            }
+        }
+    }
+
+    fun createRequest(method: String, url: String, tokens: RestOauth1AccessToken, sourceType: String): Request {
+        val authConfig = configMap[sourceType]
+                ?: throw HttpBadRequestException("client-config-not-found", "Cannot find client configurations for source-type ${sourceType}")
+        var params = this.getAuthParams(authConfig, tokens.token, tokens.tokenVerifier)
+        val signature = OauthSignature(url, params, method, authConfig.clientSecret, tokens.tokenSecret).getEncodedSignature()
+        params[OAUTH_SIGNATURE] = signature
+        val headers = mapToHeaderFormattedList(params)
+
+        val req: Request = Request.Builder()
+                .url(url)
+                .header("Authorization", "OAuth $headers")
+                .method(method, if (method == "POST") RequestBody.create(null, "") else null)
+                .build()
+        return req
+
+    }
+
+    private fun getAuthParams(authConfig: RestSourceClient, accessToken: String?, tokenVerifier: String?): MutableMap<String, String?> {
+       return mutableMapOf<String, String?>(
+            OAUTH_CONSUMER_KEY to authConfig.clientId,
+            OAUTH_NONCE to this.generateNonce(),
+            OAUTH_SIGNATURE_METHOD to OAUTH_SIGNATURE_METHOD_VALUE,
+            OAUTH_TIMESTAMP to Instant.now().epochSecond.toString(),
+            OAUTH_ACCESS_TOKEN to accessToken,
+            OAUTH_VERIFIER to tokenVerifier,
+            OAUTH_VERSION to OAUTH_VERSION_VALUE)
+    }
+
+    fun generateNonce(): String {
+        return Math.floor(Math.random() * 1000000000).toInt().toString();
+    }
+
+    fun parseParams(input: String): String? {
+        var params = input
+        params = params.replace("=".toRegex(), "\":\"")
+        params = params.replace("&".toRegex(), "\",\"")
+        return "{\"$params\"}"
+    }
+
+    fun mapToOauth2(tokens: RestOauth1AccessToken, sourceType: String): RestOauth2AccessToken {
+        // This maps the OAuth1 properties to OAuth2 for backwards compatibility in the repository
+        // Also, an additional request for getting the external ID is made here to pull the external id
+        return RestOauth2AccessToken(tokens.token, tokens.tokenSecret, Integer.MAX_VALUE,"", getExternalId(tokens, sourceType))
+    }
```

Comment:
Kotlin extension functions are a useful tool to map one object to another:
```suggestion
    fun RestOauth1AccessToken.toOAuth2(sourceType: String): RestOauth2AccessToken {
        // This maps the OAuth1 properties to OAuth2 for backwards compatibility in the repository
        // Also, an additional request for getting the external ID is made here to pull the external id
        return RestOauth2AccessToken(token, tokenSecret, Integer.MAX_VALUE,"", getExternalId(sourceType))
    }
```
Even abstract methods can be made into extension functions, e.g.
```
abstract fun RestOauth1AccessToken.getExternalId(sourceType: String): String?
```